### ISSUE TITLE
Hardware wallet integration

### ISSUE THREAD

**Issue by lukasschor** [gnosis/safe#158]

**Comment by lukasschor** [comment]
[https://app.zenhub.com/files/156711728/9a4e2f8a-3c10-451f-8471-bd1092f26182/download](https://app.zenhub.com/files/156711728/9a4e2f8a-3c10-451f-8471-bd1092f26182/download)

**Comment by lukasschor** [comment]
Integrations for first version:
- Trezor
- Ledger
- MetaMask
- Chrome Extension

**Comment by biocom** [comment]
Created a UX flow using the Status Keycard here:
https://hw2fa.netlify.com/

Depending on browser, it may need a bit to initialise the image initially.

**Comment by tschubotz** [comment]
Nice flow. My browser was struggling with the 2 links though. I downloaded the image and inspected in Preview :)

Here's my feedback so far:
![image](https://user-images.githubusercontent.com/1380529/59699346-b9b05e00-91f1-11e9-9e31-ebf82e6cc13c.png)
This text ("Lost or changed ..") does not fit to what 2FA is meant for. I would rather write something like "All transaction have to be approved by this phone and you 2FA device."

![image](https://user-images.githubusercontent.com/1380529/59699802-7aced800-91f2-11e9-8937-cc2147b00593.png)
What if the phone does not support NFC? We could grey out the status card or simple not show it. I'm in favor of somehow still telling the user why they cannot select it.

![image](https://user-images.githubusercontent.com/1380529/59699920-ae116700-91f2-11e9-9547-51bf8606aee7.png)
No, that's not possible.

![image](https://user-images.githubusercontent.com/1380529/59699933-b49fde80-91f2-11e9-9c37-431b2eee7413.png)
Should we let them confirm these?

![image](https://user-images.githubusercontent.com/1380529/59699986-d00ae980-91f2-11e9-8c16-e7267f1d4ef0.png)
If the user interrupts this by taking the card away we could write something like "Keep holding keycard" or something like this. Maybe even during the process to give some feedback. We could also work with a progress bar. 

![image](https://user-images.githubusercontent.com/1380529/59700108-fc266a80-91f2-11e9-8647-35667af77b01.png)
Do we need this? I feel like it's no necessary.

**Comment by biocom** [comment]
Also I think when tapping on SETUP KEYCARD the user would see the modal asking to hold their keycard to their phone.

**Comment by posthnikova** [comment]
This error should be everywhere where you hold the card to a phone (while identifying initially too)
http://joxi.ru/p27RGLeiKDw0pm

**Comment by biocom** [comment]
Updated the flow for when creating a new Safe:
![2FA_createSafe.png](https://images.zenhubusercontent.com/5c3f06b5116ade5d16d9b8c2/a1fccacb-5bde-48b3-8a86-332c23ae1ffd)

Changelog:
-  Replaced the 'timeout' flow with a flow where the card is held against the phone, but for a too short time span. I think this was mentioned in our last meeting.
- All hardware 2FA wallets the app supports, are now shown to the user. In case a certain HW is not supported by the user's device itself, it will be greyed out and a small note is shown why that is (e.g. requires bluetooth/nfc support).
- Added a CANCEL button on the 'Hold the keycard to your phone' screen, as requested. I don't think it's needed (and actually clutters) a dialog like the upfront PIN input.
- The 'Free slots' check is now moved sooner in the flow. Essentially we ask the user to identify the keycard -> check if it is initialised -> Yes, initialised? -> Check free slots -> No free slots? -> Show a dialog with a message explaining that. To my understanding for now we won't do the slot/card management in that flow.

2FA Transaction flow:
![2FA_transaction.png](https://images.zenhubusercontent.com/5c3f06b5116ade5d16d9b8c2/c5b36007-02b0-4662-992e-b7641f7a1632)

2FA Replace flow:
![2FA_replace.png](https://images.zenhubusercontent.com/5c3f06b5116ade5d16d9b8c2/89f1486b-ae5c-4007-a0c1-b0ef225f6179)

2FA Enable flow:
![2FA_enable.png](https://images.zenhubusercontent.com/5c3f06b5116ade5d16d9b8c2/f9e5d7c3-fa48-4457-94be-e75e2b8b9573)

Notes on the new screens:
- I specifically left out similar screens/flows we already have seen in the 'Create Safe/Setup 2FA' flow for simplicity. But obviously they will be part of the complete flow.
- On the replace 2FA screen, I propose to move the 'input recovery phrase' step before picking a new 2FA device. I think by asking it upfront:
  - The flow directly verifies if the user has permission to even replace 2FA (and thus pick a 2FA device).
  - When the user then also successfully pairs/identifies a new 2FA device later in the flow, in a way the 'friction' of having to input the recovery phrase already has been eliminated and it could be perceived as a more pleasant flow. Curious on your thoughts.

**Comment by tschubotz** [comment]
Thanks for providing these screens. Here's my feedback:

![image](https://user-images.githubusercontent.com/1380529/60024332-e18f3e00-9697-11e9-9261-38b5c44c66b9.png)
Same feedback like before: This text ("Lost or changed ..") does not fit to what 2FA is meant for. I would rather write something like "All transaction have to be approved by this phone and you 2FA device."

![image](https://user-images.githubusercontent.com/1380529/60024557-3e8af400-9698-11e9-91a3-6e50632d5e96.png)
That's the first time we work with numbers. I like that, however shouldn't these numbers then appear on the respective screens again?

![image](https://user-images.githubusercontent.com/1380529/60024600-5498b480-9698-11e9-9c2c-e42187be47c8.png)
Share does not really look tappable. Is this normal on Android?

![image](https://user-images.githubusercontent.com/1380529/60024732-9c1f4080-9698-11e9-884c-10189667f29e.png)
What if NFC is not enabled?

![image](https://user-images.githubusercontent.com/1380529/60024763-af321080-9698-11e9-86a4-ae6232adc2e6.png)
This image is used when the user should hold the keycard to their device. For a second I was confused and thought I should hold the card to the phone again. Missed the small check mark.

![image](https://user-images.githubusercontent.com/1380529/60025040-1ea80000-9699-11e9-897f-628a4d244517.png)
this is another card confirmed icon. I wonder if we should align them.

> Do we want a flow where the user can remove (not replace) a 2FA device?

**Comment by posthnikova** [comment]
> On the replace 2FA screen, I propose to move the 'input recovery phrase' step before 
> picking a new 2FA device

@biocom I agree because if the user doesn't have recovery phrase they would be frustrated in the middle of the process. 

> Error validation handling happens in the dialog, allowing the validation errors to be 
> displayed agnostic of whatever screen/view they are presented on

On iOS we don't have custom popups so all dialogs will happen on separate screens.

**Comment by biocom** [comment]
@tschubotz 
- Text: Haven't updated the placeholder text. Will update it to your suggested text.

> That's the first time we work with numbers. I like that, however shouldn't these numbers then appear on the respective screens again?

We have been using the numbers here https://zpl.io/VDBK9Wq but in a more plain text version. I don't think showing the numbers throughout the screens will benefit the user: The context of what '1' or '2' are, will be lost. Unless we use a numbered stepper for the flows, which I don't think is appropriate in this particular case. I think the main purpose of the numbered list, is to visually indicate what the steps will be going forward, so the user can anticipate on them.

> Share does not really look tappable. Is this normal on Android?

Don't know about normal: Currently we have a share button like here https://zpl.io/am6Z9wm (outside the card) and also a version like here https://zpl.io/2vlDlmn (inside the card). I think when creating the screen I found two buttons to be competing. But I do agree that it could stand out as an action. What I'll do is turn SHARE into a button, but without the primary color (e.g. white with green text).

> What if NFC is not enabled?

You'd be prompted in a dialog, after tapping the **START**  button, that NFC is not enabled. See an existing screen for this in the flows I shared:

<img width="238" alt="Screenshot 2019-06-25 at 10 55 01" src="https://user-images.githubusercontent.com/31534717/60084417-c70ca180-9737-11e9-8182-7acd9b13e53b.png">

> This image is used when the user should hold the keycard to their device. For a second I was confused and thought I should hold the card to the phone again. Missed the small check mark. 

The darker grey image (not final design FYI) was meant as purely card interaction confirmation feedback. Whereas the other one a final confirmation that the card has been setup. But I can see the confusion and would propose to consolidate the icons.

@tschubotz As for the recovery phrase order: So I understand we don't want to change this order, correct?

@posthnikova 
> On iOS we don't have custom popups so all dialogs will happen on separate screens.

Prior to creating the dialogs I did a brief research on how iOS handles this. Some things I found:
- NFC in-app tag reading (bottom dialog) -> https://developer.apple.com/design/human-interface-guidelines/ios/user-interaction/near-field-communication/
- An example implementation, which I wonder if is a custom implementation: https://blog.usejournal.com/swift-corenfc-as-quick-as-possible-79544796cd0a
![1_QHcvReA68asH18xBxFzNcw](https://user-images.githubusercontent.com/31534717/60085122-07b8ea80-9739-11e9-9626-820c885d76f9.gif)

Can you verify how this works? Because in my opinion that would provide the best experience, not having to go away from the existing view.

**Comment by posthnikova** [comment]
> Can you verify how this works? Because in my opinion that would provide the best 
> experience, not having to go away from the existing view.

@biocom Documentation for ios 13 is preliminary so it's hard to say if it's possible not to show this bottom dialog. 

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/fc689d4c-b6d5-4920-b339-ea97a101ef71)

If we use this system popup all scanning and success/error messages will be in this popup. So designs will differ from android.

**Comment by biocom** [comment]
Updated all the flows and created actual UI screens from all the UX flow presented screens:

**Create Safe:**

![2019_UI_2FA_createSafe](https://user-images.githubusercontent.com/31534717/60185235-d66c1780-9829-11e9-9d74-6703eb137e57.png)

---

**Enable 2FA:**

![2019_UI_2FA_enable](https://user-images.githubusercontent.com/31534717/60185078-94db6c80-9829-11e9-963d-183cb08e2d32.png)

---

**Disable 2FA:**

![2019_UI_2FA_disable](https://user-images.githubusercontent.com/31534717/60185097-9e64d480-9829-11e9-9ef0-83cf4fd635b0.png)

---

**Replace 2FA:**

![2019_UI_2FA_replace](https://user-images.githubusercontent.com/31534717/60185104-a1f85b80-9829-11e9-9ee1-0c286d410866.png)

---

**2FA Transaction confirmation:**

![2019_UI_2FA_transaction](https://user-images.githubusercontent.com/31534717/60185115-a886d300-9829-11e9-9d62-49c1f2902a52.png)


I will also add each individual screens to InVision.
@lukasschor We'll also need screen names, once these screens are approved and I upload them to Zeplin.

**Comment by rmeissner** [comment]
Just some edge case questions:

If we have a second factor that we don't now (e.g. we have no address to 2fa type mapping) what should happen?
Should we always fallback to the push flow (extension flow) or should we let the user decide?
This is also interesting for the case when we implement the feature, since we for existing safes that setup 2fa before this feature we don't have the 2fa type information (since we don't store the owners in the safe app)

**Comment by rmeissner** [comment]
I like the designs.

I assume @biocom left out some error and pairing flows on purpose for now?

**Comment by rmeissner** [comment]
@biocom there are also some libs for android that tell you where your NFC spot is: https://github.com/cotechde/nfc-sweetspot maybe we could make use of that ... I realized for example that our huawei has the NFC spot far more up than my Pixel

**Comment by tschubotz** [comment]
I like the designs!

> there are also some libs for android that tell you where your NFC spot is: https://github.com/cotechde/nfc-sweetspot maybe we could make use of that ... I realized for example that our huawei has the NFC spot far more up than my Pixel

I like the idea. I would say this is not needed for v1 and should / could be part of a future iteration. :)

![image](https://user-images.githubusercontent.com/1380529/60248019-1088f800-98c2-11e9-9f2c-4260d71bd99d.png)
There is a typo in succesSfully

> If we have a second factor that we don't now (e.g. we have no address to 2fa type mapping) what should happen?
> Should we always fallback to the push flow (extension flow) or should we let the user decide?
This is also interesting for the case when we implement the feature, since we for existing safes that setup 2fa before this feature we don't have the 2fa type information (since we don't store the owners in the safe app)

It feels like it's the most pragmatic solution to work with the "fallback to the push flow" solution. And still start to build up this mapping between address and 2fa type. I'm wondering if this will come back at us at some point in the future 🤔 
Because for new versions, i.e. we know, stored and maintained the mapping, when then a new owner address appeared which we don't know, it's a sign that they tampered with their Safe. In that case I would rather show some kind of error and tell them "unrecognized owner address please replace your 2FA".

**Comment by KristinaMayman** [comment]
Following user tests, added rephrasing suggestions to these screens in the comments: 
https://projects.invisionapp.com/share/D6SSBAOBJAG#/screens/371755957
https://projects.invisionapp.com/share/D6SSBAOBJAG#/screens/371755941
https://projects.invisionapp.com/share/D6SSBAOBJAG#/screens/371778869

**Comment by posthnikova** [comment]
Changes after user tests are up for review. Prototype for settings up 2FA is here: https://invis.io/UDTCH0LV3QN. What's been done:

## Create Safe: 

https://www.dropbox.com/s/2xkjpk2dw5qbkp1/Setup_2FA.jpg?dl=0

1) Onboarding - I added explanation of what Safe is on the first slide and "You're in control" part from "How it works" screen on the second slide. 

2) CreateSafe (three steps) replaces "How it works" screen. I propose an indicator which shows how many steps were completed and how many are left. This is for better understanding of where the users are in the process. It affects how "Connect Authenticator" and "Recovery phrase" will look during Safe creation. It doesn't show during password creation because we don't know if the user chooses to create or recover Safe.  

3) Setup2FA - there is now explanation of what 2FA is.

4) Setup_2FA (2FA devices list) - added information about every option and a link to learn more. 

5) Status_Keycard (intro) - now there's an explanation of what steps there are to pair the Keycard. There was a distinction between setting up the card, identifying and pairing. Now there's only pairing and activation. Happy case looks like this: hold the card to the phone, provide PIN and password, then hold the card to the phone again. When card hadn't been activated (=set up) before, it is activated automatically. There's no need to enter credentials provided on Status_Keycard (activated) screen, after that pairing happens automatically. 

6) Status_Keycard (activated) - tapping on credentials copies them to clipboard

7) Status_Keycard (success) - added explanation how to confirm future transactions, and a mention that replacing and removing 2FA are possible. 

## Send:

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/02086165-0709-4c0e-a025-24d772f0b9a0)

Not much changed, only added a reminder that blocking and restoring card are also possible during this process.

## Replace 2FA:

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/15047488-6f97-4d66-b883-917205155c88)

Changes to 2FA devices list, Status_Keycard (intro) screen and pairing process are the same as with Safe creation. 

## Enable 2FA:

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/48e18db7-240a-4ca6-845a-37ad61309fef)

Changes to 2FA devices list, Status_Keycard (intro) screen and pairing process are the same as with Safe creation. 

## Disable 2FA:

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/d73b5fd7-2421-4f64-a757-8cf758ea6179)

Added a mention that it's possible to enable 2FA in the future. Added a reminder to have recovery phrase ready.

**Comment by posthnikova** [comment]
Changes are up for review. Prototype is here: https://invis.io/UDTCH0LV3QN. What's been done:

## Create Safe:

https://www.dropbox.com/s/rmnqututn3j9hhn/Create.jpg?dl=0

Added final icons, 2 more slides to onboarding, Skip button in onboarding, text changes. I propose showing animation in the popup to make it clear that the user should hold the card to the phone: 

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/afaf2d52-0b00-478b-b0b4-97fd5294453f)

## Send:

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/58f0f59e-5f01-4202-a208-0d520f23abe9)

Added a part when PIN is incorrect. I propose showing animation in the card to attract attention to the need to confirm with 2FA. Ideally, we should detect the card on this screen: 

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/85ffb7ec-6bd7-484c-bf52-6c80110e7a2d)

## Replace 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/0000e4cc-335c-418f-afa2-536cfd34db5c)

## Enable 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/d83f1c4a-80e3-44fb-a70e-094c52486482)

## Disable 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/b603f65f-f92d-478e-98a0-f0fb9990651a)

**Comment by rmeissner** [comment]
I like the flows, but have some questions/ comments for the transaction confirmation case:
1.) Before we can say that a pin is valid/invalid the user needs to hold the card to the device
2.) To we really want that the user resets the PIN while in the sending funds flow? Would it make sense to have a separate section in the app menu for card management?
3.) For the animation (very minor): I would probably animate the card behind the phone, since for most android phone it actually doesn't work if you would put the card on top of the phone. (Else I like the animation :) )
4.) The skip button in the Safe creation flow feels a little misplaced (would move it to the bottom)
5.) For the 3 step progress, what speaks against making it for steps so that the user knows when he is really finished?
6.) More a question for @tschubotz maybe: Do you expect anything special to happen on a technical level for the "remove paired device" dialog? (e.g. remove pairing on backend, delete pairing on keycard)
7.) Personally I feel a little uncomfortable initializing the card with credentials before displaying them to users. This sounds like a setup where a user will contact us because we bricked their status card.
8.) Should we have an option to display the password on input?
9.) When should we show the keycard blocked dialog?
10.) (Maybe future feature, and I think we shortly talked about it: Do we want to offer the possibility to store the pin, so that it is not required for future interaction?)
11.) We should be able to go to the "success" screen right after the card has been initially setup.

**Comment by rmeissner** [comment]
Some technical questions:
- When using a card that has already been paired, I assume we should skip as much as possible from the flow?
- Should we use a different private key each time the card is paired to a Safe?

**Comment by tschubotz** [comment]
> 1.) Before we can say that a pin is valid/invalid the user needs to hold the card to the device

True, that part is missing.

> 2.) To we really want that the user resets the PIN while in the sending funds flow? Would it make sense to have a separate section in the app menu for card management?

Good point, I would move it to the settings to keep the flow less complicated. Let's discuss in our meeting.

> 6.) More a question for @tschubotz maybe: Do you expect anything special to happen on a technical level for the "remove paired device" dialog? (e.g. remove pairing on backend, delete pairing on keycard)

What do you think makes sense? I think both things you mentioned would make sense. What if the keycard is used on more than 1 Safe on the same device?

> Should we use a different private key each time the card is paired to a Safe?

Yes, that would be great! Also when the card is used for more than 1 Safe on the same device, if possible.

**Comment by posthnikova** [comment]
Final version is up for review. 

Individual screens Android: https://invis.io/UDTCH0LV3QN, iOS: https://invis.io/N3TNZ0TD8RZ

What's been done: 
- Text changes
- There's PIN confirmation after activating card
- Onboarding 2FA - added success screen for connecting Authenticator
- When keycard is blocked we ask to get in touch with us
- Send_Review2FAConfirmed - added rejected state
- Animations, different animation for Android. I noted where there's animation and where static image
- On onboarding slides Next button is now aligned between iOS/Android
- Pairing card starts with asking PIN and password. In case the user doesn't have them, there is a button "I don't have a password:
- Status_Keycard (cancel) - Remove button is now red
- Added recovery flow
- Checked alignment of titles and button text between iOS/Android

#Android

##Create Safe

https://www.dropbox.com/s/6fwg3786rpiivdt/Create_Android_2019-08-28.png?dl=0

##Send

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/98691411-1565-49df-9297-96698dde234d)

##Replace 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/deb4ce03-384b-4bb6-b894-325827d3073c)

##Disable 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/17dc9431-60a9-44e2-b4b5-1f9f87eb2ec3)

##Enable 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/b5f6e51b-145a-4667-b785-1e06669fe6ed)

##Recover Safe

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/8de7a5cd-47d4-4a3c-8f1d-f349673de4c0)

##Animation

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/b85c6a99-b20c-4da6-9d17-ab9c3ddf0896)

#iOS

##Create Safe

https://www.dropbox.com/s/xovocpo1zzmlwf5/Create2FA_iOS.png?dl=0

##Send

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/68aeb556-70b8-4e3f-87f7-f33dbbffe99c)

##Replace 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/40aea67a-6560-4517-9f1c-5ecd9cb5db4a)

##Disable 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/d53dcb03-a0ac-4c12-b753-59f22ec0efc7)

##Enable 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/b7093686-69b2-4757-8f7c-8cdc38876ca5)

##Recover Safe

https://www.dropbox.com/s/tzcxvgfn1kar256/Recover_iOS.png?dl=0

##Animation

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/a825836e-c00b-46ba-8cd6-14b0bd0143d8)

**Comment by DmitryBespalov** [comment]
- In the Send flow, the PIN request screens must be before the scanning of the keycard - they are after the scanning right now

Other than that, all fine by me

**Comment by rmeissner** [comment]
Still find the initialisation flow questionable. Didn't we decided that the user should first see the information that we use to setup the card, before we initialise it?

**Comment by posthnikova** [comment]
> Still find the initialisation flow questionable. Didn't we decided that the user should first see 
> the information that we use to setup the card, before we initialise it?

@rmeissner By the time the user taps "I do not have a password", they should be aware that the card would be initialised. We could rewrite the text here https://invis.io/UDTCH0LV3QN#/377680638_Status_Keycard_-intro-_ ("If you don’t have the PIN and pairing password, your Keycard should be initialised. You can proceed to initialising on the next screen.") and rename "I do not have a password" button to "Initialise card". I think we don't need a separate step to ask their consent.

**Comment by posthnikova** [comment]
> In the Send flow, the PIN request screens must be before the scanning of the keycard - 
> they are after the scanning right now

@DmitryBespalov Scanning the card first was added after the meeting. I'm confused now

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/56136f01-6083-4db6-89cc-4668fc00e931)

**Comment by rmeissner** [comment]
@posthnikova there is a difference, if we init the card and the user presses back without storing the credentials, his card becomes unusable

**Comment by posthnikova** [comment]
@rmeissner Could we remove the paired device in case they tap Back (like when canceling)? http://joxi.net/YmEyOkeIwVVPbm So they could start new pairing.

**Comment by rmeissner** [comment]
@posthnikova unpairing will not help, the user has an initialized card and didn't write down the passphrase so he cannot use the card anymore. 

@posthnikova @tschubotz @DmitryBespalov @sche @elgatovital I feel like some technical details are not clear yet and I would propose to have another short meeting to finalize this (as I feel github is very inefficient right now)

**Comment by posthnikova** [comment]
Final version is here and on zeplin. 

# Android

## Create Safe

https://www.dropbox.com/s/uf3yqzy9kr3s0pj/Android_Setup_2FA_2019-09-10.png?dl=0
https://app.zeplin.io/project/5af063c9b4dc859b6bdaf897/dashboard?seid=5d767406270a8c32e572baef

## Send

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/9ba9d32c-4163-47f3-8f46-4a65077791f9)
https://app.zeplin.io/project/5af063c9b4dc859b6bdaf897/dashboard?seid=5d767967c07a87625747787a

## Replace 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/10b1168c-ab4b-492c-a4b4-7c130c1087eb)
https://app.zeplin.io/project/5af063c9b4dc859b6bdaf897/dashboard?seid=5d767b65a4678c188d2f68f0

## Disable 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/fdf7af10-4560-4d2a-9916-05d4c941b0ca)
https://app.zeplin.io/project/5af063c9b4dc859b6bdaf897/dashboard?seid=5d767cd1e83ac45f043a32bf

## Enable 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/60466938-0040-4c6f-b582-b91e169ef6a9)
https://app.zeplin.io/project/5af063c9b4dc859b6bdaf897/dashboard?seid=5d767c79141c091a497c3c98

## Recover Safe

https://www.dropbox.com/s/dnf5ilrnoel9ybm/Android_Recover_Safe_2019-09-10.png?dl=0 
https://app.zeplin.io/project/5af063c9b4dc859b6bdaf897/dashboard?seid=5cd0415a9b7e1c03aee7e3bc

## Changes to menu

With 2FA: https://zpl.io/bl6LgO0
Without 2FA: https://zpl.io/aRXGZRK

# iOS

## Create Safe

https://www.dropbox.com/s/xgtbsg4v45d2jv1/iOS_Setup_2FA_2019-09-10.png?dl=0 
https://app.zeplin.io/project/5c49ce9866e3b3bee966f431/dashboard?seid=5d7667cd141c091a497bb488

## Send

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/ec100846-32fb-433e-abb1-005c5e150370)
https://app.zeplin.io/project/5c49ce9866e3b3bee966f431/dashboard?seid=5d766d8a0ac06bb6ea824206

## Replace 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/b2ec093a-9b1f-431b-b267-3604b6c7c807)
https://app.zeplin.io/project/5c49ce9866e3b3bee966f431/dashboard?seid=5d767013b526ed1a8602bbe1

## Disable 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/5e3de725-2c3c-458b-8b5c-bdcc421d2b5f)
https://app.zeplin.io/project/5c49ce9866e3b3bee966f431/dashboard?seid=5d767129e83ac45f0439d248

## Enable 2FA

![image.png](https://images.zenhubusercontent.com/5c3f0bad81fb545d1bd58531/beb9624e-a743-4cfd-b9d8-ee39ced702e0)
https://app.zeplin.io/project/5c49ce9866e3b3bee966f431/dashboard?seid=5d7670e068f73118f9908106

## Recover Safe

https://www.dropbox.com/s/7j0rwduxgtmplqz/iOS_Recover_Safe_2019-09-10.png?dl=0 
https://app.zeplin.io/project/5c49ce9866e3b3bee966f431/dashboard?seid=5ced475978761e1b1b13c3f5

## Changes to menu

With 2FA: https://zpl.io/bP1g79l
Without 2FA: https://zpl.io/2vP0k05
